### PR TITLE
Add captive portal functionality, update examples and documentation, resolves #134

### LINKED ISSUE #134: Make redirection of all 'not found' URLs a configuration option
Similar to discussion in issue #25 "Captive portal" I am trying to use ESPUI in a situation where I would like it to be hard for a user to fail to hit the page.

My suggestion is to make a configuration method to change the handling of 404 errors by ESPUI so OPTIONALLY they redirect to /

I have done this myself by replacing line 1135 of ESPUI.cpp...

server->onNotFound([](AsyncWebServerRequest* request) { request->send(404); });

with...

server->onNotFound([](AsyncWebServerRequest* request) { request->redirect("/"); });

but this is obviously a one-off hack, ideally it should be optionally set using a new method.

If you like this idea I might look at doing a pull request for it myself. It seems to help with the ESPUI interface being brought up automatically in captive portal type situations.

### PR DESCRIPTION
**Pull request by ncmreynolds**
I have made a small change to optionally redirect all 404s to the 'root' of the ESP, turning it into a captive portal if acting as a Wi-Fi hotspot and DNS server.

This PR includes changes to the examples and documentation to reflect this. Sorry for the long wait in my producing a complete solution to #134 I got absentminded and forgot about it until I worked on another project with ESPUI.  The existing examples can already act as Wi-Fi hotspot and DNS server so the changes to them are one line.

### CODE REVIEW

**Comment by s00500** [comment]
hey @ncmreynolds, thanks for the PR
Just one comment: Why make this off by default ? I would argue that we can always have this redirect :-)

**Comment by ncmreynolds** [comment]
I left it off default because it is probably not very useful when the ESP is acting as a station not a hotspot. Having said that having it on by default most likely wouldn't have any negative effect. Feel free to change this if you wish.

**Comment by s00500** [comment]
Please make the change, then lets merge it :-)

**Comment by ncmreynolds** [comment]
Done (I think, I'm not great with git)

File: examples/gui-generic-api/gui-generic-api.ino [hunk 1]
```diff
@@ -199,7 +199,7 @@ void setup(void)
         if (WiFi.status() != WL_CONNECTED)
         {
             Serial.print("\n\nCreating hotspot");
-
+            ESPUI.captivePortal = true; //Configure ESPUI to be a captive portal only if the ESP is acting as a hotspot
```
Comment (s00500) [inline examples/gui-generic-api/gui-generic-api.ino:202]: can be removed again

File: examples/gui/gui.ino [hunk 2]
```diff
@@ -189,7 +189,7 @@ void setup(void)
         if (WiFi.status() != WL_CONNECTED)
         {
             Serial.print("\n\nCreating hotspot");
-
+            ESPUI.captivePortal = true; //Configure ESPUI to be a captive portal only if the ESP is acting as a hotspot
```
Comment (s00500) [inline examples/gui/gui.ino:192]: can be removed again

File: examples/tabbedGui/tabbedGui.ino [hunk 3]
```diff
@@ -197,7 +197,7 @@ void setup(void)
         if (WiFi.status() != WL_CONNECTED)
         {
             Serial.print("\n\nCreating hotspot");
-
+            ESPUI.captivePortal = true; //Configure ESPUI to be a captive portal only if the ESP is acting as a hotspot
```
Comment (s00500) [inline examples/tabbedGui/tabbedGui.ino:200]: can be removed again

File: keywords.txt [hunk 4]
```diff
@@ -25,6 +25,7 @@ beginLITTLEFS	KEYWORD2
 print	KEYWORD2
 updateSwitcher	KEYWORD2
 updateSlider	KEYWORD2
+captivePortal	KEYWORD2
```
Comment (s00500) [inline keywords.txt:28]: Hm... keyword2 is used for functions... not necessarily for globals

File: src/ESPUI.cpp [hunk 5]
```diff
@@ -1351,7 +1351,16 @@ void ESPUIClass::beginLITTLEFS(const char* _title, const char* username, const c
         request->send(200, "text/plain", String(ESP.getFreeHeap()) + " In LITTLEFS mode");
     });
 
-    server->onNotFound([](AsyncWebServerRequest* request) { request->send(404); });
+    server->onNotFound([this](AsyncWebServerRequest* request) {
+		if(captivePortal == true)
```
Comment (s00500) [inline src/ESPUI.cpp:1355]: no need to check for == true, just do `if(captivePortal)`

**Review by s00500 — CHANGES_REQUESTED** [review]
Cool, still got some minor comments, sorry for nitpicking =P

**Comment by ncmreynolds** [comment]
Suggestions done, also updated the documentation change to reflect this.

**Comment by s00500** [comment]
Great, thanks a lot :-)

**Review by s00500 — APPROVED** [review]

**Pull request merged**